src/picom.c

```diff
@@ -1382,16 +1398,58 @@ static void unredirect(session_t *ps) {
 	free(ps->damage_ring);
 	ps->damage_ring = ps->damage = NULL;
 
+	if (ps->present_event_id) {
+		xcb_present_select_input(ps->c, ps->present_event_id,
+		                         session_get_target_window(ps), 0);
+		ps->present_event_id = XCB_NONE;
+		xcb_unregister_for_special_event(ps->c, ps->present_event);
+		ps->present_event = NULL;
+	}
+
 	// Must call XSync() here
 	x_sync(ps->c);
 
 	ps->redirected = false;
 	log_debug("Screen unredirected.");
 }
 
+static void handle_present_events(session_t *ps) {
+	if (!ps->present_event) {
+		return;
+	}
+	xcb_present_generic_event_t *ev;
+	while ((ev = (void *)xcb_poll_for_special_event(ps->c, ps->present_event))) {
+		if (ev->event != ps->present_event_id) {
+			continue;
+		}
+		if (ev->evtype == XCB_PRESENT_EVENT_COMPLETE_NOTIFY) {
```

Comment:
could this if also include if above? i mean
```c
...
if (ev->event == ps->present_event_id && ev->evtype == XCB_PRESENT_EVENT_COMPLETE_NOTIFY) {
...
```

Reply:
i updated the comments, hopefully it's clearer now.